### ISSUE TITLE
Replace httpretty with responses

### ISSUE THREAD

**Issue by davidread** [ckan/ckan#4755]
## tl;dr
* I want to replace the annoying 'httpretty' library with 'responses'. 
* It is used in a few of CKAN's tests, to mock out HTTP requests.
* 'responses' seems a reasonable drop-in replacement, with more support.

## Reasons to replace httpretty
It mocks by messing with socket connections. It's a pain to diagnose problems due to seemingly unrelated HTTP requests timing out or failing in weird ways.

httpretty has given us lots of grief in the past that we've done dirty hacks for:
* https://github.com/ckan/ckan/commit/7b83c947d9ef56e7cf3aba802ba4259817ac7866
* https://github.com/ckan/ckan/issues/3502

And there are some other things:
* Multiple issues on this topic has been [closed prematurely](https://github.com/gabrielfalcao/HTTPretty/issues/65#issuecomment-125938665) by the lib maintainer.
* It doesn't have official python3 support yet
* It has a dependency on urllib3 that is in [conflict with requests](https://github.com/ckan/ckan/issues/4537) which is [causing](https://github.com/ckan/ckan/pull/4539) [trouble](https://github.com/ckan/ckan/pull/4538) upgrading.

This week I hit a fresh issue again, wasting a couple of hours tracking down the problem. For the record, httpretty.activate() causes requests to SOLR to timeout upon socket closure. We identified that before with [python 2.6](https://github.com/ckan/ckan/commit/7b83c947d9ef56e7cf3aba802ba4259817ac7866), but it this time it was python 2.7.

## Alternatives

I think the main ones are: responses, httmock, mocket. Comparisons are here:

* https://python.libhunt.com/httpretty-alternatives
* https://python.libhunt.com/compare-responses-vs-httpretty?rel=cmp-lib

Responses has the most stars by a decent margin. It's clear that httpretty only has lots of github stars from when it was popular, but now Google Trends shows responses is way more popular.

'Responses' benefits:
* it is python3 compatible
* it intercepts via `requests`, rather than the socket, which seems simpler
* it is drop-in replacement for httproxy - i.e. the same basic syntax

**Comment by davidread** [comment]
CC @frafra 

@amercader I think we discussed doing this ages ago but no-one got round to it. Shall I go ahead and do a PR for it?

**Comment by davidread** [comment]
Oh it's not quite a drop-in, it captures all requests, so you need to tell it which URLs to pass-through, which usually means `responses.add_passthru('http://127.0.0.1:8983/solr')`

**Comment by amercader** [comment]
@davidread Big +1 from me, even if requires a bit of work updating syntax. I'll be happy to see httpretty gone and if it doesn't support py3 even more
